assets/src/edit-story/components/checklist/checks/videoOptimization.js

```diff
@@ -0,0 +1,215 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { _n, __ } from '@web-stories-wp/i18n';
+import styled from 'styled-components';
+
+/**
+ * Internal dependencies
+ */
+import { useCallback, useMemo, useRef } from 'react';
+import { useFeature } from 'flagged';
+import { useLocalMedia, useStory } from '../../../app';
+import { VIDEO_SIZE_THRESHOLD } from '../../../app/media/utils/useFFmpeg';
+import { PRIORITY_COPY } from '../constants';
+import { LayerThumbnail, Thumbnail, THUMBNAIL_TYPES } from '../../thumbnail';
+import {
+  CARD_TYPE,
+  ChecklistCard,
+  DefaultFooterText,
+} from '../../checklistCard';
+import { filterStoryElements } from '../utils/filterStoryElements';
+import {
+  Button,
+  BUTTON_SIZES,
+  BUTTON_TYPES,
+  Tooltip,
+} from '../../../../design-system';
+import { useHighlights } from '../../../app/highlights';
+import { useRegisterCheck } from '../checkCountContext';
+import { StyledVideoOptimizationIcon } from '../../checklistCard/styles';
+
+const STATES = {
+  UPLOADING: 'UPLOADING',
+  UPLOADED: 'UPLOADED',
+};
+
+const OptimizeButton = styled(Button)`
+  margin-top: 4px;
+`;
+
+export function isVideoElementOptimized(element = {}) {
+  if (element.resource?.isTranscoding) {
+    return true;
+  }
+
+  if (
+    element.type !== 'video' ||
+    element.resource?.local ||
+    element.resource?.isOptimized
+  ) {
+    return false;
+  }
+
+  const videoArea =
+    (element.resource?.height ?? 0) * (element.resource?.width ?? 0);
+  const isLargeVideo =
+    videoArea >= VIDEO_SIZE_THRESHOLD.WIDTH * VIDEO_SIZE_THRESHOLD.HEIGHT;
+  return isLargeVideo;
+}
+
+export const BulkVideoOptimization = () => {
+  const enableBulkVideoOptimization = useFeature('enableBulkVideoOptimization');
+
+  const story = useStory(({ state }) => state);
+  const optimizingResourcesRef = useRef({});
+  const unoptimizedVideos = filterStoryElements(story, isVideoElementOptimized);
+
+  const setHighlights = useHighlights(({ setHighlights }) => setHighlights);
+
+  const { optimizeVideo } = useLocalMedia((state) => ({
+    optimizeVideo: state.actions.optimizeVideo,
+  }));
+
+  const sequencedVideoOptimization = useCallback(() => {
+    // only attempt to optimize videos that are not currently being transcoded
+    const optimizingResources = unoptimizedVideos.filter(
+      ({ resource }) => !resource?.isTranscoding
+    );
+    return new Promise((res) => {
+      optimizingResources
+        .map(({ resource }) => async () => {
+          await optimizeVideo({ resource }).finally(() => {
+            optimizingResourcesRef.current[resource.id] = STATES.UPLOADED;
+          });
+        })
+        .reduce((p, fn) => p.then(fn), Promise.resolve())
+        .then(() => {
+          res();
+        });
+    });
```

Comment:
```suggestion
    return new Promise((resolve, reject) => {
      optimizingResources
        .map(({ resource }) => async () => {
          await optimizeVideo({ resource }).finally(() => {
            optimizingResourcesRef.current[resource.id] = STATES.UPLOADED;
          });
        })
        .reduce((p, fn) => p.then(fn), reject())
        .then(resolve);
    });
```

Would this work?

Reply:
Would if the `optimizeVideo` fails? Does there need to be a catch?

Reply:
I'll test